### ISSUE TITLE
Create OCR for existing assets

### ISSUE THREAD

**Issue by jrochkind** [sciencehistory/scihist_digicoll#2306]
Using code from https://github.com/sciencehistory/scihist_digicoll/pull/2302

With works to include spec'd in #2080 

This is proving more troublesome than expected, so I'm creating a ticket for it especially.  For tracking and cost-estimating. 

- [x] Rake task run in production
- [x] Jobs completed in production

**Comment by jrochkind** [comment]
This is taking longer than I thought -- meaning more heroku charges than I thought for our worker dynos. 

In my past experience, I thought OCR was going to be around 3-5 seconds a page. And for some pages it is. 

But others can be 10, 20, or even 40 seconds. 

This may be because the tasks are very CPU intensive, so competing for CPU on our `standard-2x` dynos, which share CPU cores with other users. 

Using a performance dyno isn't necessarily more cost-effective, since they don't have a lot of cores either. Even performance-l only has 8 cores, even though we can fit many more than 8 workers on one, they'll be fighting for cpu. Possibly better, or the same, to spread them across standard-2x's. 

Once we get things working, we will be able to estimate total cost.

**Comment by jrochkind** [comment]
However, at present things aren't even working.....

Oddly, while tesseract works on _staging_, on _production_, we are getting this error when trying to run tesseract:

--------------
Aug 08 15:39:43 app/special_worker.4 WARN: [ActiveJob] [CreateAssetOcrJob] [83c5ec80-113a-4e99-b769-cdc5a1a6146f] Retrying CreateAssetOcrJob (Job ID: 83c5ec80-113a-4e99-b769-cdc5a1a6146f) after 1 attempts in 0 seconds, due to a TTY::Command::ExitError (Running `tesseract -c textonly_pdf\=1 -c tessedit_page_number\=0 /tmp/shrine20230808-1564-9fgoox.tif /tmp/scihist_digicol_asset_hocr_creator20230808-1564-vxst8h/tesseract_out -l eng hocr pdf` failed with
Aug 08 15:39:43 app/special_worker.4   exit status: 1
Aug 08 15:39:43 app/special_worker.4   stdout: Nothing written
Aug 08 15:39:43 app/special_worker.4   stderr: Error opening data file /usr/share/tesseract-ocr/4.00/tessdata/eng.traineddata
Aug 08 15:39:43 app/special_worker.4 Please make sure the TESSDATA_PREFIX environment variable is set to your "tessdata" directory.
Aug 08 15:39:43 app/special_worker.4 Failed loading language 'eng'
Aug 08 15:39:43 app/special_worker.4 Tesseract couldn't load any languages!
Aug 08 15:39:43 app/special_worker.4 Could not initialize tesseract.
----------------

Since staging and production should be identical, this is very odd.

**Comment by jrochkind** [comment]
OK.....   tesseract files do not end up installed in ordinary places, so we DO need to set `TESSDATA_PREFIX`. 

And it was set on staging. 

    TESSDATA_PREFIX:                  /app/.apt/usr/share/tesseract-ocr/4.00/tessdata

We need to set it on production too. We also need to document.  I probably originally got the solution from https://towardsdatascience.com/deploy-python-tesseract-ocr-on-heroku-bbcc39391a8d

It actually already was documented! at https://sciencehistory.atlassian.net/wiki/spaces/HDC/pages/2213052417/Installing+tesseract

Will edit to add more, and more links. 

Also will add tesseract to `system_env_tests`.

**Comment by jrochkind** [comment]
Cost estimate:

We have 25,000 assets to OCR under spec. 

Running 14 `standard-2x` dynos as special workers. standard-2x cost about $.069/hour, so 14 of them are $0.966 an hour (just under a dollar an hour). 

When I sampled, 14 workers processed about 82 assets a minute, so 4920 an hour. So it would take about 5.08 hours to process, or $4.90.

**Comment by CHF-IT-Staff** [comment]
Thanks Jonathan

**Comment by jrochkind** [comment]
OK, most have completed, but htere are 20 "incorrigible" ones that fail every time they are tried. Listing by UUID pk since that's what I get from the logs:

* 7235b719-8cc5-4274-9722-e97d1cf3fcf2
* e72c3d4d-a39c-4607-8edd-c6449f0dee5c
* 1f799a99-9c1b-43f7-a1b8-2e08e5a6ab0b
* ecb44240-5cd0-4ae5-b22e-7875e0eebb14
* 550a5553-e769-4bcd-b8cf-b97b9ccc4eec
* 8d730e3c-f05f-4b31-ba00-7854ee57dbba
* dde5bb22-c2c9-48d5-b90a-3882832fa8be
* d867f7df-6ca0-4883-9460-19780a92a49e
* 311045de-c719-4073-810d-d99fe55389dc
* 02a6a7bf-0a58-46af-8efd-3ca738cd1a9e
* da61247a-2da6-460b-b545-38f7877d3d57
* d8cb3295-72e1-4f89-8b41-bd6de9ef1ff3
* 12ca919f-2cdd-4f13-a2f9-07ce39dc3223
* feb68068-c01c-4df0-a915-0223291f03bf
* cf452424-8763-4489-9ca5-df49d05019d0
* d81b2eb0-513f-4933-8a84-2ff8a6873d11
* 3dc417ba-6dec-4a16-a976-aa796edfef17
* cdb1a2d4-b602-4d1b-b723-90a0d5820643
* 78515fcb-7f6e-4416-a565-7d563607d30c

My guess is that for most of these, tesseract ends up using a lot of RAM trying ot OCR them, more than is available on heroku -- either heroku kills the processes, or they are so slow using swap eventually they time out. 

Will investigate each individually for salient characteristics. 

We could try temporarily using a special work dyno that has a lot more RAM to see if it does better, a performance-m (2.5GB) or even performance-l (14GB).  While these dynos are expensive if left up, if we only run them for an hour or two it's no big deal, about 35 cents or 70 cents an hour.

**Comment by eddierubeiz** [comment]
Interesting...

**Comment by jrochkind** [comment]
OK, I have a spreadsheet with info I'm trying to figure out how to share a link to here. In the meantime...

* One of them is a corrupt file: https://digital.sciencehistory.org/admin/asset_files/m1zy3an

* Around 8 of them are VERY LARGE photographs, which do not have any actual text, and do not need OCR:
   * example https://digital.sciencehistory.org/admin/asset_files/7crmftg
   * Most of these are actually single-image child works, that we could turn off "OCR enabled"  for the work (as long as we still have the child work feature), but....

* A couple are huge stitched-together images, that don't _neccessarily_ need to get OCR'd (some have no text, some have limited text), but are in works with other pages that probably could, eg https://digital.sciencehistory.org/admin/asset_files/eft6bg1

* A couple are giant stitched together newspaper pages with LOTS of text, that are clearly just too much for tesseract, or use too much RAM for our heroku dynos -- even if it would be nice to have OCR. I think these two: https://digital.sciencehistory.org/admin/works/lo68iua#tab=nav-members. We _could_ turn off OCR for that work

* A few are complete mysteries, they seem like ordinary pages to me, why are THESE giving us trouble? I may try them again to verify. Eg. https://digital.sciencehistory.org/admin/asset_files/ws859g75h and https://digital.sciencehistory.org/admin/asset_files/3484zj040 (both of those are from the same work)

We could simply say those ~20 assets aren't going to get OCR'd -- the PDFs generated just won't have text layers, they won't be included in a site search, etc, no big deal, it's just a few outliers. 

BUT as it is, the nightly job will run every night and TRY to OCR them and fail, which we don't want. 

For SOME of them we could just turn off OCR at the work-level, but at least a few of them are pages in works that otherwise should get OCR'd. 

I can think of two solutions to the "infinite loop of trying to OCR known problem images"

1. We introduce a new asset-level toggle, so we can toggle them off. Perhaps ti's just a "suppress OCR" toggle -- the Work toggle is still primary, but you can "opt out" individual assets. (You can't opt them in without the work level being enrolled). 

2. We record errors somewhere, and the nightly job (or the WorkOcrCreatorRemover in general) know NOT to try to OCR an asset which has an OCR error recorded.  With a way to "erase" the error if you do want to try again. 

I'm not sure which makes more sense, or is easier. Maybe #1 is easier/simpler, on the development end? But requires a develoepr to notice something is error'ing a lot, and manually dis-enroll it from OCR. 

Other ideas?

Separately, we could try to figure out WHY these aren't OCR'ing. 

If it's possible to tesseract them on a MacOS with more RAM etc, we could figure out a good way to do that and then attach both the `hocr` file and the `textonly_pdf` file to an Asset in production, in a way that is easy and minimizes any manual errors (when working with files in production!). Not sure if we'd need to generally add UI to admin to let people manually attach these derivatives (a pain) or if we just create some kind of command line to do it (also kind of a pain to figure out how to get files to production).

**Comment by jrochkind** [comment]
Spreadsheet of Assets that currently can't be OCR'd is at:

`P://Support/Computer Services/Digital Collections/External Access/assets_failed_ocr.xlsx`

https://files.sciencehistory.org/dl/pmpR1uDbcA/assets_failed_ocr.xlsx_

I note admin URLs to each, and also additional info I have figured out about each.

**Comment by jrochkind** [comment]
@apinkney0696 says

> An asset-level OCR toggle sounds preferable to me. There are going to be plenty of works for which not every page is going to need OCR, (I'm thinking illustrations, foldouts, randomly poor OCR pages, etc). It will be nice to have the ability to suppress asset OCR on a case-by-case basis.

> Additionally, I think it could be worth figuring out the workflow for "manual OCR" eventually, aka using Tesseract on a personal device and then uploading the hocr via command line. However, I don't think we should spend time on that now unless it makes sense to tackle in the context of your other current work.

**Comment by jrochkind** [comment]
OK, trying to run tesseract on some of these through tesseract on my Mac. The first one I tried _did_ succeed!

https://digital.sciencehistory.org/admin/works/lo68iua

I ran it like this on my Mac to measure time/memory of the exact thing we're running:

>  command time -l tesseract -c textonly_pdf=1 -c tessedit_page_number=0 ~/Downloads/americas_enemies_are_lo68iua_9_fjkxzwr.tiff tess_lo68iua -l eng hocr pdf

(Output of "maximum resident set size" is in _bytes_, and is approximate RAM consumption). 

That succeeded in 68 seconds, taking about 1.3GB of RAM.  

I will run on a few others, and note in spreadsheet.

**Comment by jrochkind** [comment]
OK, four of the mystery failures that didn't seem unusual or large -- I was able to succesfully OCR (on heroku infrastructure) trying them again manually. I guess they were just bad luck temporary failure somehow -- they are now marked as such in spreadsheet. 

I took a sampling of other failures that actually had text to OCR, and ran them through tesseract on my Mac. They all succeeded on Mac -- but taking from 328MB to up to 1.5GB of RAM.   So that's the limit that is causing the failure. 

Most of them are unusually large files.  Only one seems like a typical file (158MB, 6526 x 8243 pixels), but still took a huge 752MB of RAM on my Mac -- https://digital.sciencehistory.org/admin/asset_files/8940pz9 . So that one is a little bit mysterious _why_ it took so much RAM -- but universally the cause for the remaining failures is "taking more RAM than avail on our heroku dynos).
